### PR TITLE
Add CpuSurfaceUploadFilter test to ImageBlit suite to explicitly test linear filter pipeline

### PR DESCRIPTION
**Pull request by stianeklund**
Add CpuSurfaceUploadFilter test to ImageBlit suite

Tests the round-trip fidelity of CPU-written surface data through xemu's internal scale path.

Specifically this targets the bug which is fixed in [xemu-project/xemu#2874](https://github.com/xemu-project/xemu/pull/2874), where `pgraph_vk_upload_surface_data` used the linear filter instead of nearest causing blurred visuals when upscaling CPU-written surfaces to the internal GPU resolution.

The test writes 1-pixel-wide alternating red/blue vertical stripes directly to the framebuffer via CPU, then issues a minimal GPU draw to trigger the surface upload. Reading the framebuffer back via CPU then triggers the download path. With Vulkan as the renderer and 1x (no upscale) the round-trip should be pixel-exact.

This test will fail with current xemu 0.8.135 with the Vulkan renderer with an upscale >= 2x (but passes with the fixes mentioned above).

Resulting pass for PR branch mentioned earlier:
xemu-2026-05-24-18-37-44

Current xemu 0.8.135:

xemu-2026-05-24-18-38-25

Just for posterity, I ran this on real hardware, too (it should pass), this is through a HDMI converter -> HDMI capture device so it's already upscaling a bit: 
Screenshot 2026-05-24 19-10-33

Please let me know if you want any changes, I am really over my head with this already so I do want to emphasize that I actually know very little about this stuff, but I'm willing to learn.

### CODE REVIEW

File: src/tests/image_blit_tests.cpp [hunk 1]
```diff
@@ -185,6 +186,13 @@ static std::string MakeTestName(uint32_t clip_x, uint32_t clip_y, uint32_t clip_
  * @tc Overlap_BR_Inside
  *   Tests the boundary condition of xemu's 2D blit overlap detection by performing a 1x1 pixel blit exactly inside the
  *   bottom-right corner of a 3D render surface.
+ *
+ * @tc CpuSurfaceUploadFilter
+ *   Writes 1-pixel-wide alternating R/B vertical stripes to the framebuffer via CPU, then issues a GPU draw to
+ *   trigger xemu's internal surface upload (pgraph_vk_upload_surface_data). After GPU completion the framebuffer
+ *   is read back via CPU, which triggers xemu's download path. At scale > 1x the NEAREST+NEAREST round-trip
+ *   preserves exact R/B values, while LINEAR+NEAREST (the pre-fix behaviour) produces blended intermediate values
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: nit: Let's change the latter part of this to "while any other filtering produces blended intermediate values ..." alternatively, you could file a bug against xemu capturing the misbehavior and then reference it here.

File: src/tests/image_blit_tests.cpp [hunk 2]
```diff
@@ -33,6 +33,7 @@ static constexpr char kBlitRenderBlitTest[] = "BlitRenderBlit";
 static constexpr char kOverlapFIFOTest[] = "OverlapFIFO";
 #endif
 static constexpr char kBlitPastWidthTest[] = "BlitBeyondWidth";
+static constexpr char kCpuSurfaceUploadFilterTest[] = "CpuSurfaceUploadFilter";
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: Let's rename this to `XemuScaledSurfaceUploadFilter` since the test is specific to xemu's extension of Xbox behavior and requires render scaling to be enabled. We'll also need to make some changes to the test harness to facilitate this; currently we only run tests at 1x.

File: src/tests/image_blit_tests.cpp [hunk 3]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  static constexpr uint32_t kColorA = 0x00FF0000;  // red
+  static constexpr uint32_t kColorB = 0x000000FF;  // blue
+  // NEAREST+NEAREST round-trips to 0xFF/0x00; LINEAR upload + NEAREST download blends to ~0x80.
+  static constexpr uint8_t kChannelThreshold = 0xE0;
+
+  host_.PrepareDraw(0xFF111111);
+
+  auto *fb = static_cast<uint32_t *>(pb_agp_access(pb_back_buffer()));
+  const uint32_t pitch_pixels = pb_back_buffer_pitch() / 4;
+  const uint32_t stripe_x = (host_.GetFramebufferWidth() - kTestW) / 2;
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: nit: I think "start_x" and "start_y" would be clearer since this is calculating the offsets for each column and row.

File: src/tests/image_blit_tests.cpp [hunk 4]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  host_.SetDiffuse(0.f, 0.f, 0.f, 0.f);
+  host_.SetScreenVertex(630.f, 460.f, 0.f);
+  host_.SetScreenVertex(634.f, 460.f, 0.f);
+  host_.SetScreenVertex(634.f, 464.f, 0.f);
+  host_.SetScreenVertex(630.f, 464.f, 0.f);
+  host_.End();
+
+  host_.PBKitBusyWait();
+
+  fb = static_cast<uint32_t *>(pb_agp_access(pb_back_buffer()));
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: I don't think there's any need to re-calculate this, `fb` isn't modified.

File: src/tests/image_blit_tests.cpp [hunk 5]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  host_.PrepareDraw(0xFF111111);
+
+  auto *fb = static_cast<uint32_t *>(pb_agp_access(pb_back_buffer()));
+  const uint32_t pitch_pixels = pb_back_buffer_pitch() / 4;
+  const uint32_t stripe_x = (host_.GetFramebufferWidth() - kTestW) / 2;
+  const uint32_t stripe_y = (host_.GetFramebufferHeight() - kTestH) / 2;
+
+  for (uint32_t y = 0; y < kTestH; ++y) {
+    for (uint32_t x = 0; x < kTestW; ++x) {
+      fb[(stripe_y + y) * pitch_pixels + (stripe_x + x)] = (x % 2 == 0) ? kColorA : kColorB;
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: nit: On a slow CPU recalculating the Y offset for each X could add up to meaningful time. I'd move the stripe_y ... calculation out of the inner loop. You could also premultiply `stripe_y` and just add `pitch_pixels` to save a multiplication.

I suspect Clang's optimizer would already do this for release builds, but if there's some reason to debug w/ optimization off, this could be meaningful.

File: src/tests/image_blit_tests.cpp [hunk 6]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+
+  bool pass = true;
+  uint32_t fail_count = 0;
+  uint32_t first_actual = 0;
+  uint32_t first_expected = 0;
+  static constexpr uint32_t kMaxLoggedFailures = 16;
+
+  for (uint32_t y = 0; y < kTestH; ++y) {
+    for (uint32_t x = 0; x < kTestW; ++x) {
+      const uint32_t actual = fb[(stripe_y + y) * pitch_pixels + (stripe_x + x)];
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: Same nit as above, consider moving the Y calculation and multiplication out of the inner loop.

File: src/tests/image_blit_tests.cpp [hunk 7]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+      const uint8_t actual_r = (actual >> 16) & 0xFF;
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: Can you walk me through the logic here? We're accessing whole pixels that were populated precisely via the CPU, but then doing a fuzzy check on the red and blue channels. Is the threshold needed in practice?

`NEAREST` should snap to an actual existing pixel value, no?

File: src/tests/image_blit_tests.cpp [hunk 8]
```diff
@@ -837,6 +846,86 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+      const bool expect_red = (x % 2 == 0);
+
+      const bool pixel_pass = expect_red
+          ? (actual_r >= kChannelThreshold && actual_b <= (0xFF - kChannelThreshold))
+          : (actual_b >= kChannelThreshold && actual_r <= (0xFF - kChannelThreshold));
+
+      if (!pixel_pass) {
+        if (fail_count < kMaxLoggedFailures) {
+          PrintMsg("FAIL [%u,%u]: expected %s, got 0x%06X\n", x, y,
+                   expect_red ? "RED" : "BLU", actual & 0xFFFFFF);
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp]: nit: "BLUE" (though generally the tests don't bother with debug messages for failures, visibly different output is sufficient)

**Comment by abaire** [comment]
Thanks!

FYI - the tests already save the framebuffer to the HDD for you, so you should be able to run this on HW and retrieve results without going through any HDMI/etc... encoding.

File: src/tests/image_blit_tests.cpp (hunk 7 above)
Comment (stianeklund) [reply src/tests/image_blit_tests.cpp]: Yes, it should, this isn't needed. It was added defensively, the threshold AFAIK shouldn't be necessary at all. The idea was to try to detect blending failure without requiring exact values. This is also left over from the first implementation I did, will fix.

**Comment by stianeklund** [comment]
@abaire thanks for the review. I've addressed your comments. I checked the HDD but I can't actually see any saved output, does it save attempt to save to `D:\` if run from that partition?

File: src/tests/image_blit_tests.cpp (hunk 2 above)
Comment (stianeklund) [reply src/tests/image_blit_tests.cpp]: Good input, done. I'm curious what would be necessary to achieve this though? Is there any guest side control of render scale in xemu? I guess you could in theory use two separate configs or separate CI jobs..? (assuming the config exposes the render scale setting). I do think that's a bit outside of scope of this PR though but I do see the value in having perhaps a xemu-specific test harness for sure.

Comment (abaire) [reply src/tests/image_blit_tests.cpp]: Agreed, it's entirely outside of the scope of this change.

https://github.com/abaire/nxdk-pgraph-test-runner and the CI scripts for https://github.com/abaire/xemu-nxdk_pgraph_tests_results and https://github.com/abaire/xemu-dev_pgraph_test_results will need to be updated to produce a config with >1x render scale, likely along with a test set filtered to contain just this test initially.

We may want to update this repo to make it easier to enumerate all tests along with some metadata so we can tag tests that have xemu-specific requirements alongside the test; it'd be difficult to discover this cross-repo requirement otherwise.

**Comment by abaire** [comment]
> @abaire thanks for the review. I've addressed your comments. I checked the HDD but I can't actually see any saved output, does it save attempt to save to `D:\` if run from that partition?

https://github.com/abaire/nxdk_pgraph_tests/blob/16c2782712d92136903d804576ceb091305bbd46/CMakeLists.txt#L88 if you haven't overridden it, `E:\nxdk_pgraph_tests\`

**Comment by abaire** [comment]
Looks like clang-format needs to be run; you can also set up the pre-commit hook in githooks to do this automatically on commit.

If you've already run it and it's still different, let me know. Unfortunately clang-format isn't versioned so it's possible that your local version and the CI version are out of sync.

**Comment by stianeklund** [comment]
Please do not merge this until the respective PR in xemu is merged I think (if it will be) otherwise I'm not sure I see the value in it, unless you're OK with it explicitly failing.

**Comment by abaire** [comment]
Sounds good, I'll keep an eye on your xemu PR; I agree with your perspective that it doesn't make sense to add a xemu-specific test for something that may be considered by-design behavior.

**Comment by stianeklund** [comment]
@abaire xemu PR got merged,are you OK with merging this or do you want any further changes?

**Comment by abaire** [comment]
Let me catch up with the xemu PR and I'll take another look at this. I should have time in the next day or two.

If you could grab the raw framebuffer cap off your HW and have it ready to create a PR against https://github.com/abaire/nxdk_pgraph_tests_golden_results that'd also be helpful.

File: src/tests/image_blit_tests.cpp [hunk 9]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
   FinishDraw(kBlitRenderBlitTest);
 }
 
+void ImageBlitTests::TestXemuScaledSurfaceUploadFilter() {
+  static constexpr uint32_t kTestW = 128;
+  static constexpr uint32_t kTestH = 128;
+  static constexpr uint32_t kColorA = 0x00FF0000;  // red
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:852]: Super nitpick: if it matters that they're red and blue, just name them kColorRed and kColorBlue? If it doesn't matter, let's remove the comments in case somebody finds a reason to change the colors in the future.

Given that you use "expect_red" below, I think calling it `kColorRed`/`Blue` would be appropriate, otherwise I'd rename the boolean to match (e.g., `expect_color_a`).

File: src/tests/image_blit_tests.cpp [hunk 10]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+    for (uint32_t x = 0; x < kTestW; ++x) {
+      fb[row_offset + x] = (x % 2 == 0) ? kColorA : kColorB;
+    }
+  }
+
+  // Minimal GPU draw to trigger the surface upload and set draw_dirty.
+  // Placed in the bottom-right corner to avoid the test region.
+  host_.SetFinalCombiner0Just(TestHost::SRC_DIFFUSE);
+  host_.SetFinalCombiner1Just(TestHost::SRC_ZERO, true, true);
+  host_.Begin(TestHost::PRIMITIVE_QUADS);
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:873]: Another nitpick, since this draw is irrelevant except for its side effect, I'm curious:

a) if it still works if you render something outside of the viewing frustum (e.g., at 641, 0)
b) if it has to be offscreen, using PRIMITIVE_POINTS and rendering a single dot in a corner would be less test code.

File: src/tests/image_blit_tests.cpp [hunk 11]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  uint32_t first_expected = 0;
+  static constexpr uint32_t kMaxLoggedFailures = 16;
+
+  row_offset = start_y * pitch_pixels + start_x;
+  for (uint32_t y = 0; y < kTestH; ++y, row_offset += pitch_pixels) {
+    for (uint32_t x = 0; x < kTestW; ++x) {
+      const uint32_t actual = fb[row_offset + x];
+      const bool expect_red = (x % 2 == 0);
+      const uint32_t expected = (expect_red ? kColorA : kColorB) & 0xFFFFFF;
+      const bool pixel_pass = (actual & 0xFFFFFF) == expected;
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:895]: nit: I'd just put this into if statement since it's not reused.

File: src/tests/image_blit_tests.cpp [hunk 12]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  bool pass = true;
+  uint32_t fail_count = 0;
+  uint32_t first_actual = 0;
+  uint32_t first_expected = 0;
+  static constexpr uint32_t kMaxLoggedFailures = 16;
+
+  row_offset = start_y * pitch_pixels + start_x;
+  for (uint32_t y = 0; y < kTestH; ++y, row_offset += pitch_pixels) {
+    for (uint32_t x = 0; x < kTestW; ++x) {
+      const uint32_t actual = fb[row_offset + x];
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:892]: Given that every subsequent use of `actual` masks off the high bits, consider doing that here at assignment so future changes can't accidentally forget the mask and cause errors.

File: src/tests/image_blit_tests.cpp [hunk 13]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+  host_.SetDiffuse(0.f, 0.f, 0.f, 0.f);
+  host_.SetScreenVertex(630.f, 460.f, 0.f);
+  host_.SetScreenVertex(634.f, 460.f, 0.f);
+  host_.SetScreenVertex(634.f, 464.f, 0.f);
+  host_.SetScreenVertex(630.f, 464.f, 0.f);
+  host_.End();
+
+  host_.PBKitBusyWait();
+
+  bool pass = true;
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:883]: I don't think there's a need to track pass as a separate bool, you're already counting the failures using `fail_count` and can simplify this to a check whether `fail_count` is non zero (in this you can just treat `fail_count` exactly like a bool and invert your `pass` logic).

File: src/tests/image_blit_tests.cpp [hunk 14]
```diff
@@ -837,6 +846,79 @@ void ImageBlitTests::TestBlitRenderBlit() {
...
+        }
+        pass = false;
+        ++fail_count;
+      }
+    }
+  }
+
+  pb_print("%s\n", kXemuScaledSurfaceUploadFilterTest);
+  pb_print("1px R/B stripes: CPU write -> GPU draw -> readback\n");
+  pb_print("Result: %s\n", pass ? "PASS" : "FAIL");
```
Comment (abaire) [inline src/tests/image_blit_tests.cpp:913]: to clarify the comment above: 

`fail_count ? "FAIL" : "PASS"` on this line

`if (fail_count) {` on the next line
